Added concept of test pyramid with examples how this can be applied w… (#7260)
jenkins-infra/jenkins.io: content/doc/developer/testing/index.adoc

== What to Test
Now that we can write a basic test, we should discuss what you should be testing…

The test pyramid is a concept used in software development to describe the ideal distribution of different types of automated tests that make up a test suite.
The idea was introduced by Mike Cohn and is visualized as a pyramid to help illustrate the optimal proportion of each type of test.
Here's a breakdown of the three main layers from the bottom to the top of the pyramid:

1. *Unit Tests*: At the base of the pyramid, unit tests are the most numerous.
These tests are focused on checking the smallest parts of an application, such as individual functions or methods.
Unit tests are quick to execute and help ensure that each component of the software works as expected in isolation.

2. *Integration Tests*: The middle layer of the pyramid consists of integration tests.
These are fewer in number compared to unit tests.
Integration tests verify that different parts of the system work together as intended.

> For Jenkins, this makes use of the `JenkinsRule` to test the integration of the plugin into Jenkins.

Reviewer: ```suggestion
For Jenkins, this makes use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] to test the integration of the plugin into Jenkins.
```

This can be a test for a new build step and testing this with a test pipeline.

3. *End-to-End Tests (E2E) and UI Tests*: At the top of the pyramid, end-to-end tests are the fewest in number but typically the most complex.
These tests simulate real user scenarios from start to finish.
They interact with the application as a user would, testing the complete flow of the system.

> For Jenkins, this makes use of the `JenkinsRule` or `JenkinsSessionRule` in combination with the `createWebClient()`-call which is used to really creates a call to endpoint to assert something on the result.

Reviewer: ```suggestion
For Jenkins, this makes use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] or link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsSessionRule.html[`JenkinsSessionRule`] in combination with the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html#createWebClient()[`createWebClient()`]-call which is used to really creates a call to endpoint to assert something on the result.
```

> Purpose of the Test Pyramid: The pyramid aims to encourage developers to write a larger number of lower-level tests (unit tests) and fewer high-level tests (E2E tests). This distribution is recommended because lower-level tests tend to be faster, cheaper to automate, and more reliable, whereas higher-level tests are slower, more expensive, and can be flakier.

Reviewer: We've preferred to use sentence per line for easier formatting.  Optional change, not required.
```suggestion
Purpose of the Test Pyramid: The pyramid aims to encourage developers to write a larger number of lower-level tests (unit tests) and fewer high-level tests (E2E tests).
This distribution is recommended because lower-level tests tend to be faster, cheaper to automate, and more reliable, whereas higher-level tests are slower, more expensive, and can be flakier.
```

1. *Unit Tests*: At the base of the pyramid, unit tests are the most numerous.
These tests are focused on checking the smallest parts of an application, such as individual functions or methods.
Unit tests are quick to execute and help ensure that each component of the software works as expected in isolation.

2. *Integration Tests*: The middle layer of the pyramid consists of integration tests.
These are fewer in number compared to unit tests.
Integration tests verify that different parts of the system work together as intended.
For Jenkins, this makes use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] to test the integration of the plugin into Jenkins.
This can be a test for a new build step and testing this with a test pipeline.

3. *End-to-End Tests (E2E) and UI Tests*: At the top of the pyramid, end-to-end tests are the fewest in number but typically the most complex.
These tests simulate real user scenarios from start to finish.
They interact with the application as a user would, testing the complete flow of the system.

> For Jenkins, this makes use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] or link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsSessionRule.html[`JenkinsSessionRule`] in combination with the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html#createWebClient()[`createWebClient()`]-call which is used to really creates a call to endpoint to assert something on the result.

Reviewer 2: All tests that use the `jenkins-test-harness` are still integration tests and *not* End-to-End Tests. End-to-End tests require a real browser to be used and can be accomplished only by using the acceptance test harness. You can find some real End-to-End test examples in the warnings and git-forensics plugins (the idea and initial implementation is from @olivergondza).

> For Jenkins, this makes use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] or link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsSessionRule.html[`JenkinsSessionRule`] in combination with the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html#createWebClient()[`createWebClient()`]-call which is used to really creates a call to endpoint to assert something on the result.

Reviewer 3: ```suggestion
For Jenkins, this makes use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] or link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsSessionRule.html[`JenkinsSessionRule`] in combination with the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html#createWebClient()[`createWebClient()`]-call, which is used to create a call to an endpoint to then assert something on the result.
```
Small suggestion to adjust the wording of the last line here, as its current syntax is not clear.

The test pyramid is a concept used in software development to describe the ideal distribution of different types of automated tests that make up a test suite.
The idea was introduced by Mike Cohn and is visualized as a pyramid to help illustrate the optimal proportion of each type of test.
Here's a breakdown of the three main layers from the bottom to the top of the pyramid:

1. *Unit Tests*: At the base of the pyramid, unit tests are the most numerous.
These tests are focused on checking the smallest parts of an application, such as individual functions or methods.
Unit tests are quick to execute and help ensure that each component of the software works as expected in isolation.

2. *Integration Tests*: The middle layer of the pyramid consists of integration tests.
These are fewer in number compared to unit tests.
Integration tests verify that different parts of the system work together as intended.
For Jenkins, this makes use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] to test the integration of the plugin into Jenkins.

> This can be a test for a new build step and testing this with a test pipeline.
> 
> 3. *End-to-End Tests (E2E) and UI Tests*: At the top of the pyramid, end-to-end tests are the fewest in number but typically the most complex.
> These tests simulate real user scenarios from start to finish.
> They interact with the application as a user would, testing the complete flow of the system.
> For Jenkins, integration tests including the UI, make use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] or link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsSessionRule.html[`JenkinsSessionRule`] in combination with the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html#createWebClient()[`createWebClient()`]-call, which is used to create a call to an endpoint to then assert something on the result.
> The End-To-End Tests are done via the link:https://github.com/jenkinsci/acceptance-test-harness[`acceptance-test-harness`] or can be done with a real browser, as can be seen in the link:https://github.com/jenkinsci/warnings-ng-plugin/tree/main/ui-tests[warnings-ng ui-tests].

Reviewer 2: I would move the HTML unit tests to topic 2 ( but maybe I am too pedantic as I am teaching Mike Cohns definition in my testing courses 🙈)

```suggestion
For Jenkins, integration tests including the UI make use of the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html[`JenkinsRule`] or link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsSessionRule.html[`JenkinsSessionRule`] in combination with the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/JenkinsRule.html#createWebClient()[`createWebClient()`]-call, which is used to create a call to an HTML-Unit endpoint to then assert something on the result. 

This can be a test for a new build step and testing the results in the UI with a test pipeline.

3. *End-to-End Tests (E2E) and UI Tests*: At the top of the pyramid, end-to-end tests are the fewest in number but typically the most complex.
These tests simulate real user scenarios from start to finish.
They interact with the application as a user would, testing the complete flow of the system.
The End-To-End Tests are done with a real browser in combination with the link:https://github.com/jenkinsci/acceptance-test-harness[`acceptance-test-harness`], as can be seen in the link:https://github.com/jenkinsci/warnings-ng-plugin/tree/main/ui-tests[warnings-ng ui-tests].

These tests are typically not used in Jenkins plugins because they are too complex to create and maintain.
```
